### PR TITLE
Add nextfloat(::BigFloat, n), prevfloat(::BigFloat,n) 

### PR DESCRIPTION
**Pull request by narendrakpatel**
* New constructor for BigFloat. refer [this](https://github.com/JuliaLang/julia/issues/31276#issuecomment-471166758)
* Improve nextfloat/prevfloat and introduce nextfloat! and prevfloat!
* Introduce nextfloat(::BigFloat, n) and prevfloat(::BigFloat, n)
* Tests for constructor, nextfloat and prevfloat

Closes: #31276

### CODE REVIEW

File: base/mpfr.jl
Comment (StefanKarpinski) [inline base/mpfr.jl]: Technically when `n == 0` we could just return `x`. Not sure it's worth the complexity though.

**Comment by StefanKarpinski** [comment]
This is looking fairly good. There's a key API question in here: do we want `BigFloat(x, new=true)` to be the official way to ask for a newly allocated big float?

File: base/mpfr.jl
Comment (narendrakpatel) [reply base/mpfr.jl]: Actually, the special case `n==0` is already considered in `nextfloat!()`. Note that `nextfloat()` had to return a copy of x, so anyhow we would need to call Bigfloat constructor. Thus no special case was specifically created for this function.

Comment (StefanKarpinski) [reply base/mpfr.jl]: > Actually, the special case `n==0` is already considered in `nextfloat!()`

But that's too late: `BigFloat(x, new=true)` has already allocated a new copy of `x` by the time `nextfloat!` gets the value.

Comment (narendrakpatel) [reply base/mpfr.jl]: My argument is that, nextfloat() needs to return a new copy of x, as it does not ends with `!`. So even in case n==0, we would need to create a copy of x. Thus, no computation is wasted.

**Comment by narendrakpatel** [comment]
We had a long discussion on what the API of the BigFloat should look like [here](https://github.com/JuliaLang/julia/issues/31276). I believe the summary of the whole discussion can be found [here](https://github.com/JuliaLang/julia/issues/31276#issuecomment-471166582). @JeffreySarnoff can elaborate more. What are your views on it?

File: base/mpfr.jl
Comment (StefanKarpinski) [reply base/mpfr.jl]: > My argument is that, `nextfloat()` needs to return a new copy of `x`

Returning the same value is not a mutating operation. Since BigFloats are considered to be immutable it's fine to return the same value instead of allocating a new identical value.

**Comment by StefanKarpinski** [comment]
That's fine, I'm calling broader attention to it to get some agreement before merging.

File: base/mpfr.jl
Comment (simonbyrne) [inline base/mpfr.jl]:
```suggestion
                    precision::Integer=MPFR.precision(x), new::Bool=precision != MPFR.precision(x))
```

File: base/mpfr.jl
Comment (simonbyrne) [inline base/mpfr.jl]:
```suggestion
    new || return x
```

Comment (narendrakpatel) [reply base/mpfr.jl]: Does this sound good?
```julia
precision != MPFR.precision(x) && !new &&
                    @warn "new=false is ignored because precision differ" &&
                    return BigFloat(x, precision=precision)
new || return x
```

Comment (narendrakpatel) [reply base/mpfr.jl]: The precision is being copied. In the constructor, default value of precision is precision(x).

Comment (StefanKarpinski) [reply base/mpfr.jl]: Ok, then with @simonbyrne's new default value I'm more ok with this.

File: base/mpfr.jl
Comment (narendrakpatel) [reply base/mpfr.jl]: Does this look good?
```julia
nextfloat(x::BigFloat, n::Int) = n == 0 ? x : nextfloat!(BigFloat(x, new=true), n)
```

File: base/mpfr.jl
Comment (JeffreySarnoff) [reply base/mpfr.jl]: me too

File: base/mpfr.jl
Comment (StefanKarpinski) [reply base/mpfr.jl]: Yes, that's just it.

File: base/mpfr.jl
Comment (JeffBezanson) [reply base/mpfr.jl]: I recall a whole discussion where we concluded that `BigFloat(x::BigFloat)` should use the current global precision, not the precision of `x`. Are we sure we want to change that? It doesn't seem like something we can just change casually as part of dealing with `nextfloat` and copying.

The `new` keyword argument seems quite ad-hoc to me; since no other functions support it it's just a more verbose way to add a function for this purpose. It also didn't occur to me before that it conflicts with the precision behavior of the constructor, which I think is a pretty significant issue.

Comment (JeffreySarnoff) [reply base/mpfr.jl]: I thought there was a later extended discussion where we concluded that using the global precision was most appropriate as a fallback, default and that ignoring the given precision of an extant `BigFloat` when copying it had been determined to be more problematic than beneficial.  If that is untrue, and I grant that it may be, then please note that much of my [advice](https://github.com/JuliaLang/julia/issues/31276) earlier was based upon that supposition.

Comment (JeffBezanson) [reply base/mpfr.jl]: I was thinking of #29127. Is there a link to the later discussion?

Comment (JeffreySarnoff) [reply base/mpfr.jl]: That's the one :blush:!

Comment (JeffreySarnoff) [reply base/mpfr.jl]: I think we need a way to not force conformance to the global setting.  If I include a package that exports `const HiPrecisionPi` which is a `BigFloat` with a precision greater than the default, should it be incumbent on me to know that I must reset the global precision in order to work with, say,
`const HiPrecisionHalfPi = HiPrecisionPi/2` at the higher precision?

Comment (narendrakpatel) [reply base/mpfr.jl]: I found a problem with this new BigFloat API. By using such implementation, `setprecision` function breaks. Consider the following example:
```julia
julia> x = BigFloat(12.12)
12.1199999999999992184029906638897955417633056640625

julia> setprecision(128) do
           y = BigFloat(x)
           y.prec
       end
256
```
I think this would cause a lot of problems as current behaviour of `setprecision` is:
```julia
julia> x = BigFloat(12.12)
12.1199999999999992184029906638897955417633056640625

julia> setprecision(128) do
           y = BigFloat(x)
           y.prec
       end
128
```

Comment (JeffreySarnoff) [reply base/mpfr.jl]: Valid. What is your suggestion?

Comment (JeffreySarnoff) [reply base/mpfr.jl]: The proper value of `nextfloat(::BigFloat)` or `prevfloat(::BigFloat)` is precision dependant.  The value obtained that defers to the global precision is not the same as the value obtained that defers to the precision of argument when that precision differs from the global precision.

It is reasonable to decide how we want to handle this.  If the determination is that `nextfloat` and `prevfloat` use the global precision then there is no reason to copy maintaining the precision of a the value copied [at least for this purpose].  If the determination is that `nextfloat` and `prevfloat` are requesting steps in units of the precision of the argument, then there is reason to have at least an internal function that copies maintaining the precision of the value copied.  In that case, we are saying something about the primacy of the precision attached to a given `BigFloat` over the global precision setting.  And that may warrant more adaption.

Comment (narendrakpatel) [reply base/mpfr.jl]: This method was previously rejected but this might solve our problems:
```julia
function BigFloat(x::BigFloat, r::MPFRRoundingMode=ROUNDING_MODE[];
                           precision::Integer=0, new::Bool=false)
      precision==MPFR.precision(x) && !new && return x
      z = BigFloat(;precision=ifesle(precision==0,
                                  ifelse(new, MPFR.precision(x), DEFAULT_PRECISION[]), 
                                  precision))
      ccall((:mpfr_set, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode),
                z, x, r)
       return z
```
Maybe we can try to iterate over this method to achieve backward compatibilty.

Comment (JeffreySarnoff) [reply base/mpfr.jl]: We should decide about using `new` as a parameter first. imo my prior note gives a way to determine that which strips away some of the dross.

Comment (JeffreySarnoff) [reply base/mpfr.jl]: _trying to avoid going in circles, and attempting top recapitulate the generally favoured facets that have arisen while not involving any controversial aspects ..._

within mpfr.jl and __not__ exported:
these functions use the value and precision of `x`
```
function nextfloat!(x::BigFloat)
    ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), x) != 0
    return x
end

function prevfloat!(x::BigFloat)
    ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), x) != 0 
    return x
end
```

in mpfr.jl and exported
these functions respect the value and precision of `x`
```
nextfloat(x::BigFloat) = nextfloat!(x)
prevfloat(x::BigFloat) = prevfloat!(x)
```

`nextfloat!(x,n), nextfloat(x,n), prevfloat!(x,n), prevfloat(x,n)` accordingly

Comment (JeffreySarnoff) [reply base/mpfr.jl]: I think this does what we want and avoids the controversial.

defined in mpfr and __not__ exported, these respect the precision of `x`
```
function nextfloat!(x::BigFloat)
    ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), x) != 0
    return x
end

function prevfloat!(x::BigFloat)
    ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), x) != 0 
   return x
end
```

defined in mpfr.jl and exported, these also respect the precision of `x`
without influencing any other aspects of the `BigFloat` interface.
```
nextfloat(x::BigFloat) = nextfloat!(x)
prevfloat(x::BigFloat) = prevfloat!(x)
```

`nextfloat(x,n), prevfloat(x,n), nextfloat!(x,n), prevfloat!(x,n)` accordingly

Comment (JeffreySarnoff) [reply base/mpfr.jl]: ^ revised

Comment (StefanKarpinski) [reply base/mpfr.jl]: +:100: to the non-controversial approach of introducing an unexported `duplicate` function for this PR. That let's us get this PR merged adding the desired `nextfloat` and `prevfloat` methods and we can have a separate discussion of the best API for creating copies.

Comment (JeffreySarnoff) [reply base/mpfr.jl]: @narendrakpatel please ^, do as we discussed last night.

**Comment by Keno** [comment]
We took a brief look at this during triage. Two comments: We don't like the `new` keyword argument, but that seems to have been discussed extensively already. The other observation is that this seems like it changes the rounding mode behavior of the nextfloat/prevfloat functions and is thus a breaking change.

**Comment by narendrakpatel** [comment]
@Keno If we implement the `_duplicate` like this then I think that issue will be solved.
```julia
function _duplicate(x::BigFloat)
    z = BigFloat(;precision=precision(x))
    ccall((:mpfr_set, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode),
            z, x, ROUNDING_MODE[])
    return z
end
```
Basically only change in proposed implementation would be the use of `z=BigFloat(x, precision=precision(x))` instead of `z=BigFloat(x)` in nextfloat/prevfloat functions.
Shall I move forward with this approach?

**Comment by JeffreySarnoff** [comment]
@narendrakpatel Because there is now exported `nextfloat(::BigFloat)` and `prevfloat(::BigFloat)` and they use `precision(BigFloat)` rather than the precision of their argument, to do anything other than that in implementing `nextfloat(::BigFloat, ::Int)` and `prevfloat(::BigFloat, ::Int)` would require (for logical consistancy) that the current `nextfloat` and `prevfloat` implementations change to use the precision of their arguments.

That change would be "breaking" -- so it is unavailable until Julia v2.
This means that the implementation of `nextfloat(::BigFloat, ::Int)` and `prevfloat(::BigFloat, ::Int)` must be limited to the equivalent of repeatedly calling the currently exported versions `nextfloat(::BigFloat)` and `prevfloat(::BigFloat)`.

All of the good thought and refinement that we have developed over the past days will be useful.  Either as contributing to the "reform" of BigFloats for v2, or in providing important principles for a complementary package, say `LargeFloats`.

I know others are paying attention to what is and is not "breaking".  It's most important to the viability of Julia as the user base grows.  My perspective is more about where we might be going, or how I might go where I like with Julia.

So .. given this, why not take a shot at cleanly coding the "works like repeated calls to _" versions of these functions for `BigFloat`.

**Comment by JeffreySarnoff** [comment]
Is that ^ where we are at, @Keno?

**Comment by Keno** [comment]
I'm just the scribe here ;). If `nextfloat(::BigFloat, n)` didn't work before, it can do whatever, we just can't change any behavior that worked before (if we think it's minor, we can make that decision and put it in a minor version, but that would need to be explicitly decided and called out).

**Comment by JeffreySarnoff** [comment]
As far as I know, nobody uses `prevfloat(BigFloat)`, `nextfloat(BigFloat)` -- perhaps there is someone who once did something with that -- but the way people have related to BigFloats is not one that naturally leads to thinking of their previous and next floating point values.  Perhaps the best evidence of this is that until I noticed it, there were no mentions of the lack of an nth next or nth prev.

There is a good deal of support for the PR as @narendrahkpatel is about to provide.
@narendrakpatel Now Is The Time.

**Comment by JeffreySarnoff** [comment]
Until v1.1, there was no `precision` keyword to use constructing a `BigFloat`.
There was an unnamed positional arg which would pass along a value for the precision, which was of heroically little use because as soon as on did anything numerical with a `BigFloat`, the precision of the result has been slammed with the current value of `precision(BigFloat)`.  Other than looking at digits, using the unnamed second arg for setting custom precision has not been generally helpful -- and so, has not been seen as a tool of the BigFloat realm.

That `nextfloat` and `prevfloat` stepped bits at the current value of `precision(BigFloat)` is exactly how they will behave going forward when given a BigFloat that has been constructed using the current `precision(BigFloat)`. Which is exactly as they have behaved when given a BigFloat that has been constructed using the current `precision(BigFloat)` <or any other precision -- so what? people do not use BigFloats with varying precisions in `nextfloat`, `prevfloat`>.

(a) there is no pattern of use for `nextfloat, prevfloat` with BigFloats because the question of precision has lacked resonance beyond `256 bits, that's good` maybe for this `128 bits` or for that `1024 bits` (b) the sort of numerical processing that benefits from adjusting the precision does not work very well when the number type keeps slamming the current precision onto old and new results alike.

That `nextfloat` and `prevfloat` will respect the precision of the values given means they will work just as they have worked (had they been used) as the values given had been of the current precision (had they been given).  Allowing more precise and well-mannered numerical exploration going forward is only upside.

**Comment by narendrakpatel** [comment]
@StefanKarpinski @Keno Does this PR require any other changes?

File: base/mpfr.jl
Comment (rfourquet) [inline base/mpfr.jl]: For style consistency, it seems better to add space around `==` (same for `prevfloat!`).

Comment (rfourquet) [reply base/mpfr.jl]: But actually, this line seems unnecessary, the same result is achieved by an empty `for`-loop below.

File: base/mpfr.jl
Comment (rfourquet) [inline base/mpfr.jl]: Why not instead add a default argument, like `nextfloat!(x::BigFloat, n::Integer=1)` ?

**Comment by rfourquet** [comment]
This new implementation of `nextfloat(::BigInt)` seems more useful to me, and I already considered proposing such a change: indeed it's difficult for  me to imagine a use-case for `nextfloat(x)` which doesn't have the same precision as `x`. On the other hand, the current behavior is the one which is consistent with `setprecision`, so it's not an obvious choice. Anyway, I say that just to suggest that explicit documentation of this corner case (deviation from the consistency of `setprecision`) should be added to this PR.

File: test/mpfr.jl
Comment (rfourquet) [inline test/mpfr.jl]: You could test that `nextfloat(x) == nextfloat(x, 1)`. Also can you add a test for `abs(n) > 1`?

File: base/mpfr.jl
Comment (JeffreySarnoff) [reply base/mpfr.jl]: I think that does not decomplexify; and since `nextfloat!` is being introduced and not exported .. keeping it as simple as possible lets the exported function be more reliably maintained.

File: test/mpfr.jl
Comment (JeffreySarnoff) [reply test/mpfr.jl]: and that `prevfloat(x, 0) == x`

Comment (JeffreySarnoff) [reply test/mpfr.jl]: that `nextfloat(x, -2) == prevfloat(x, 2)`

File: base/mpfr.jl
Comment (rfourquet) [reply base/mpfr.jl]: Well, I proposed this change precisely because that seems simpler to me, since default args are a well understood feature, and this allows to merge two definitions into one, with easier maintainability IMHO.

Comment (JeffreySarnoff) [reply base/mpfr.jl]: I missed some of your earlier comment.  You are talking about giving `n` the default value of `1`.  I'd have no objection to that in any of the four functions: `nextfloat!`, `prevfloat!`, `nextfloat`, `prevfloat` were it not the case that `float.jl` does not give `nextfloat`, `prevfloat` default values.  I'm sure that's at least in part because `float.jl` was written eariler.  Nonetheless, there is a proclivity to follow the conventions that are in use.

Comment (rfourquet) [reply base/mpfr.jl]: > float.jl does not give nextfloat, prevfloat default values.

My guess would be that it's because those functions are documented separately. By the way, the docstring of `nextfloat(x, n)` should be updated here to include `BigFloat` (or simply to remove the type constraint `IEEEFloat`, which is not even exported or documented).

> proclivity

nice new word learnt, thanks!

Comment (narendrakpatel) [reply base/mpfr.jl]: I just checked the code in `float.jl`. It has been implemented like this `nextfloat(x::AbstractFloat) = nextfloat(x,1)`. We can do the same thing here. Also, should I change the rounding mode to global default or use `MPFRRoundNearest` instead?

Comment (rfourquet) [reply base/mpfr.jl]: > We can do the same thing here

We can also do it differently, this is what default arguments are for.

Comment (JeffreySarnoff) [reply base/mpfr.jl]: Why is it a breaking change -- what is the rounding now?

Comment (narendrakpatel) [reply base/mpfr.jl]: Currently, we can do `setrounding(T, Mode)` and the `nexfloat()` uses that Mode. What is implemented in this PR is RoundNearest Mode. 
I would prefer the method mentioned by @rfourquet but would not like to break the code style.

Comment (JeffreySarnoff) [reply base/mpfr.jl]: Yes, however when there is no `setrounding` set, the default is `RoundNearest`.  So its not clear to me that using `RoundNearest` is breaking anything.  If `setrounding` sets rounding to `RoundDown` or to `RoundUp` what difference does it make -- the nextfloat is the same as the nextfloat rounded down and the same as the nextfloat rounded up.

**Comment by JeffreySarnoff** [comment]
We have reached a meeting of appropriateness (absence of dissent, presence of consent), right? right.  @narendrakpatel has been diligent in his incorporation of the best recentest both in test and in src.  Let's clear this PR into Base.
<or please explain why not that is better>.

File: base/mpfr.jl
Comment (rfourquet) [reply base/mpfr.jl]: > I would prefer the method mentioned by @rfourquet but would not like to break the code style.

I wonder why you would think that this would break code style. It would rather be your current version which breaks style. If there is no speed advantage (which I doubt in this case) to having two separate bodies with repetition, just merge it into one function with default argument.

**Comment by rfourquet** [comment]
> @narendrakpatel has been diligent in his incorporation of the best recentest both in test and in src. Let's clear this PR into Base.

I wish some more tests and documentation like we suggested, and ideally an answer to my comments. But more importantly, regarding the breaking change, an explicit decision should be taken...

**Comment by JeffreySarnoff** [comment]
@narendrakpatel I thought you had added those tests already -- please do.  And the docs modification too.
@rfourquet What is the breaking change -- how does the rounding mode break anything if the result of `nextfloat` `prevfloat` is isomorphic to computation under `RoundNearest` whatever the RoundingMode setting?

**Comment by rfourquet** [comment]
> What is the breaking change

See [my previous post](https://github.com/JuliaLang/julia/pull/31310#issuecomment-474927819), the current behavior of `nextfloat` is to respect `setprecision`, i.e. `nextfloat(x)` doesn't necessarily create a float with the same precision as `x`. I am personally in favor of the proposed change, but this is breaking; so it must be decided whether we want this change, and if yes, whether this is a "minor change".

**Comment by StefanKarpinski** [comment]
The issue is marked for triage so we'll discuss and decide next week. It seems kind of nonsensical for `nextfloat` and `prevfloat` to give a float value with a different precision so I'm also in favor but it is breaking so it should be discussed and would be a "minor change" at least.

**Comment by StefanKarpinski** [comment]
Triage is in favor of changing this but it's too big a change to backport.

**Comment by StefanKarpinski** [comment]
Needs a NEWS entry.

**Comment by JeffBezanson** [comment]
:+1: LGTM now.

**Comment by StefanKarpinski** [comment]
This looks good to go aside from needing NEWS 😀

**Comment by JeffreySarnoff** [comment]
supergood -- getting all newsy

**Comment by JeffreySarnoff** [comment]
supergood -- getting all newsy

**Review by rfourquet — APPROVED** [review]

**Comment by JeffreySarnoff** [comment]
There is a typo and an inoperative link in the News item:

> New methods `nexfloat(::BigFloat, n::Int)` and `prevfloat(::BigFloat, n::Int)` are
    introduced. ([#31310])   ⬳ omitted the 't' in 'nextfloat'

> > New methods `nextfloat(::BigFloat, n::Int)` and `prevfloat(::BigFloat, n::Int)` are
    introduced. ([#31310](https://github.com/JuliaLang/julia/pull/31310))

Be careful to copy the full link into News:  _in html it looks like this:_
(&lt; a href="https://github.com/JuliaLang/julia/issues/31310">#31310</a&gt;)
_to appear as this:_ (<a href="https://github.com/JuliaLang/julia/issues/31310">#31310</a>)

consider this

`nextfloat(::BigFloat)` and `prevfloat(::BigFloat)` now return a value that matches their args precision. Additionally, `nextfloat(::BigFloat, n::Int)` and `prevfloat(::BigFloat, n::Int)` have now been added.

**Review by JeffreySarnoff — APPROVED** [review]

**Comment by JeffreySarnoff** [comment]
@narendrakpatel On behalf of the other participants, thank you for this awesome __first PR__.

**Comment by narendrakpatel** [comment]
If this is okay, can this be merged? @StefanKarpinski @simonbyrne

File: NEWS.md
Comment (rfourquet) [inline NEWS.md]: nitpick, if you happen to re-push before merge: "uses" -> "use"

File: NEWS.md
Comment (rfourquet) [inline NEWS.md]: nitpick, if you happen to re-push before merge: `n::Integer`

Comment (StefanKarpinski) [reply NEWS.md]:
```suggestion
  * `nextfloat(::BigFloat, n::Integer)` and `prevfloat(::BigFloat, n:: Integer)` methods have been added
```

File: NEWS.md
Comment (StefanKarpinski) [reply NEWS.md]:
```suggestion
* `nextfloat(::BigFloat)` and `prevfloat(::BigFloat)` now return a value with the same
   precision as their argument, which means that (in particular)
   `nextfloat(prevfloat(x)) == x` whereas previously this could result in a completely
   different value with a different precision ([#31310])
```

File: NEWS.md
Comment (JeffreySarnoff) [reply NEWS.md]: @narendrakpatel is in class and will push this modification  when he gets home.

Comment (JeffreySarnoff) [reply NEWS.md]: or -- just go ahead and make the change -- we all agree

**Comment by StefanKarpinski** [comment]
Looks good modulo doc suggestions.

File: NEWS.md
Comment (narendrakpatel) [reply NEWS.md]: I am back. I will make the changes right away.

Comment (JeffreySarnoff) [reply NEWS.md]: do you know how to "squash"? I don't.

Comment (narendrakpatel) [reply NEWS.md]: For now, I will just use `git commit --amend` but you can squash commit in the following way:
* `git rebase -i HEAD~n`
* replace `pick` with `squash`

Comment (StefanKarpinski) [reply NEWS.md]: You don't have to do anything. It can be squashed by the person doing the merge.

Comment (StefanKarpinski) [reply NEWS.md]: (But yes, it can be done by you as well with the appropriate git incantations.)

**Comment by narendrakpatel** [comment]
@StefanKarpinski I made all the changes that are recommended.

File: NEWS.md
Comment (StefanKarpinski) [inline NEWS.md]:
```suggestion
  * `inv(::Missing)` has now been added and returns `missing` ([#31408]).
```
Rebase artifact? This shouldn't be changing the issue number for a different NEWS entry.

File: NEWS.md
Comment (StefanKarpinski) [inline NEWS.md]:
```suggestion
  as their argument, which means that (in particular) `nextfloat(prevfloat(x)) == x` whereas
```

File: NEWS.md
Comment (StefanKarpinski) [inline NEWS.md]:
```suggestion
  previously this could result in a completely different value with a different precision ([#31310])
```

**Comment by StefanKarpinski** [comment]
Ok, we're converging here. Now it's just the rebase artifact and markdown indentation.

File: NEWS.md
Comment (narendrakpatel) [reply NEWS.md]: Actually, I worked on that issue. And while I was adding `news` entry for the current issue, I realized that I had made a mistake in `inv` issue. That's I made the change.

File: NEWS.md
Comment (narendrakpatel) [reply NEWS.md]: It is just a markdown thing. I had `build` the docs and checked that it looks fine. Do you still want me to push to these changes?

File: NEWS.md
Comment (StefanKarpinski) [reply NEWS.md]: Ah, ok. That should in principle be done in a separate commit but probably not worth it.

File: NEWS.md
Comment (StefanKarpinski) [reply NEWS.md]: Probably not worth it, although it does match the way the rest of the file is formatted. You can also just merge the suggested changes through the GitHub UI.

Comment (narendrakpatel) [reply NEWS.md]: Changed the files.

**Comment by StefanKarpinski** [comment]
This is ready to be squash-merged as soon as tests pass. Great work, @narendrakpatel—thanks for sticking it out to the end!

**Comment by narendrakpatel** [comment]
If this is okay, can this be merged? @StefanKarpinski 
PS: some tests are failing but the reasons are not related to this issue.

**Comment by StefanKarpinski** [comment]
Exciting! Congrats on the PR, @narendrakpatel!

**Pull request merged**
